datameta/security/__init__.py

```diff
@@ -197,42 +229,74 @@ def get_password_reset_token(db: Session, token: str):
         )).one_or_none()
 
 
-def revalidate_user(request):
-    """Revalidate the currently logged in user and return the corresponding user object. On failure,
-    raise a 401"""
+def revalidate_user_token_based(request, token):
+    """Revalidates user during token-based operations."""
     db = request.dbsession
-    # Check for token based auth
-    token = get_bearer_token(request)
-    if token is not None:
-        token_hash = hash_token(token)
-        apikey = db.query(ApiKey).join(User).filter(and_(
-            ApiKey.value == token_hash,
-            User.enabled.is_(True)
-            )).one_or_none()
-        if apikey is not None:
-            if check_expiration(apikey.expires):
-                raise HTTPUnauthorized()
-            return apikey.user
+
+    token_hash = hash_token(token)
+    apikey = db.query(ApiKey).join(User).filter(and_(
+        ApiKey.value == token_hash,
+        User.enabled.is_(True)
+    )).one_or_none()
+
+    if apikey is not None:
+        apikey_expired = check_expiration(apikey.expires)
+        user = apikey.user
+
+        if apikey_expired:
+            request.tm.abort()
+            request.tm.begin()
+            register_failed_login_attempt(db, user)
+            request.tm.commit()
+            request.tm.begin()
         else:
-            raise HTTPUnauthorized()
+            log.warning(f"CLEARING FAILED LOGIN ATTEMPTS FOR APIKEY.USER {user.id}")
+            user.login_attempts.clear()
```

Comment:
Since this is done a couple times we should have a `clear_login_attempts(db_user)` function that creates the log entry and performs the clearing. I don't think we need specific log entries for the different routes of user auth. But we can add this to the security split refactoring to not make merging more difficult than necessary.